File: source/uems-agent-chat/src/tools/load-guidelines.ts
/**
 * Tool: uems_agent_load_guidelines
 * Load engineering guidelines, security standards, and coding conventions
 * from the synced guidelines store (global storage) instead of cloning the repo.
 */

import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs/promises';
import { jsonResult } from './helpers';

interface LoadGuidelinesInput {
  platform?: string;
  category?: string;
  files?: string[];
}

interface GuidelineFile {
  path: string;
  content: string;
}

export class LoadGuidelinesTool implements vscode.LanguageModelTool<LoadGuidelinesInput> {
  constructor(private readonly repoDir: string) {}

  async invoke(
    options: vscode.LanguageModelToolInvocationOptions<LoadGuidelinesInput>,
    token: vscode.CancellationToken,
  ): Promise<vscode.LanguageModelToolResult> {
    void token;
    const platform = options.input.platform ?? 'mac';
    const category = options.input.category ?? 'all';
    const fileFilter = options.input.files?.length ? new Set(options.input.files) : null;

    const guidelinesDir = path.join(this.repoDir, 'guidelines');
    const docStandardsDir = path.join(this.repoDir, 'agents', 'document-generator', 'doc-standards');
    const reviewStandardsDir = path.join(this.repoDir, 'agents', 'delta-reviewer', 'review-standards');

    const result: Record<string, GuidelineFile[]> = {};

    try {
      if (category === 'all' || category === 'common') {
        result.common = await this.readDir(path.join(guidelinesDir, 'common'), 'guidelines/common', fileFilter);
      }

      if (category === 'all' || category === 'platform') {
        const platformDir = path.join(guidelinesDir, platform);
        result.platform = await this.readDir(platformDir, `guidelines/${platform}`, fileFilter);
      }

      if (category === 'all' || category === 'doc-standards') {
        result['doc-standards'] = await this.readDir(docStandardsDir, 'agents/document-generator/doc-standards', fileFilter);
      }

      if (category === 'all' || category === 'review-standards') {
        result['review-standards'] = await this.readDir(reviewStandardsDir, 'agents/delta-reviewer/review-standards', fileFilter);
      }

      const totalFiles = Object.values(result).reduce((sum, files) => sum + files.length, 0);
      if (totalFiles === 0) {
        return jsonResult({
          error: 'No guideline files found. Run "UEMS Agent Chat: Sync Agent Files" first.',
        });
      }

      return jsonResult({ platform, category, files: result });
    } catch {
      return jsonResult({
        error: 'Guidelines not synced yet. Run "UEMS Agent Chat: Sync Agent Files" to fetch them.',
      });
    }
  }

  private async readDir(dir: string, label: string, fileFilter: Set<string> | null): Promise<GuidelineFile[]> {
    const files: GuidelineFile[] = [];
    try {
      const entries = await fs.readdir(dir, { withFileTypes: true });
      for (const entry of entries) {
        if (entry.isFile() && entry.name.endsWith('.md')) {
          if (fileFilter && !fileFilter.has(entry.name)) { continue; }
          const filePath = path.join(dir, entry.name);
          const content = await fs.readFile(filePath, 'utf-8');
          files.push({ path: `${label}/${entry.name}`, content });
        }
      }
    } catch {
      // Directory doesn't exist — skip
    }
    return files;
  }
}

File: source/uems-agent-web/backend/tools/load-guidelines.go
package tools

import (
	"context"
	"os"
	"path/filepath"
	"strings"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type loadGuidelinesInput struct {
	Platform string   `json:"platform,omitempty" jsonschema:"Platform: mac linux windows"`
	Category string   `json:"category,omitempty" jsonschema:"Category: all common platform doc-standards review-standards"`
	Files    []string `json:"files,omitempty" jsonschema:"Specific filenames to load"`
}

func registerLoadGuidelines(s *mcp.Server) {
	mcp.AddTool(s, &mcp.Tool{
		Name:        "uems_agent_load_guidelines",
		Description: "Load UEMS engineering guidelines, security standards, coding conventions, doc-standards, and review-standards from the synced store. Returns file contents organized by category. Use this instead of cloning the uems-ai-toolkit repo — guidelines are synced automatically.",
	}, func(_ context.Context, _ *mcp.CallToolRequest, input loadGuidelinesInput) (*mcp.CallToolResult, any, error) {
		if input.Platform == "" {
			input.Platform = "mac"
		}
		if input.Category == "" {
			input.Category = "all"
		}

		fileFilter := map[string]bool{}
		for _, f := range input.Files {
			fileFilter[f] = true
		}

		result := map[string][]map[string]string{}

		if input.Category == "all" || input.Category == "common" {
			dir := filepath.Join(guidelinesDir, "common")
			result["common"] = readGuidelineDir(dir, "guidelines/common", fileFilter)
		}
		if input.Category == "all" || input.Category == "platform" {
			dir := filepath.Join(guidelinesDir, input.Platform)
			result["platform"] = readGuidelineDir(dir, "guidelines/"+input.Platform, fileFilter)
		}
		if input.Category == "all" || input.Category == "doc-standards" {
			dir := filepath.Join(guidelinesDir, "..", "agents", "document-generator", "doc-standards")
			result["doc-standards"] = readGuidelineDir(dir, "agents/document-generator/doc-standards", fileFilter)
		}
		if input.Category == "all" || input.Category == "review-standards" {
			dir := filepath.Join(guidelinesDir, "..", "agents", "delta-reviewer", "review-standards")
			result["review-standards"] = readGuidelineDir(dir, "agents/delta-reviewer/review-standards", fileFilter)
		}

		total := 0
		for _, files := range result {
			total += len(files)
		}
		if total == 0 {
			return errorResult("No guideline files found. Check guidelines directory path."), nil, nil
		}

		return jsonResponse(map[string]interface{}{
			"platform": input.Platform,
			"category": input.Category,
			"files":    result,
		}), nil, nil
	})
}

func readGuidelineDir(dir, label string, filter map[string]bool) []map[string]string {
	var files []map[string]string
	entries, err := os.ReadDir(dir)
	if err != nil {
		return files
	}
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".md") {
			continue
		}
		if len(filter) > 0 && !filter[e.Name()] {
			continue
		}
		content, err := os.ReadFile(filepath.Join(dir, e.Name()))
		if err != nil {
			continue
		}
		files = append(files, map[string]string{
			"path":    label + "/" + e.Name(),
			"content": string(content),
		})
	}
	return files
}
